_posts/2016-12-28-Day_Twentyone.md

```diff
@@ -0,0 +1,197 @@
+to use macros in your code should be given serious consideration and justified
+by significant benefits.
+
+I do not like magic tricks in programming. If something is mysteriously happening, I
+get this uncomfortable feeling that I am not really in control of my code: what
+if the magical thing stops working? How will I fix it if I have no clue on how it
+operates?
+
+I am not saying that I want to understand everything that goes on in a program,
+down to electrons flowing inside the processor, the same way a car driver does
+not need to know how the engine and transmission system work.
+This is called information hiding and
+it is a great thing. What I am not happy with is magic that happens at the same
+level of abstraction that I am in. For example, the Elixir Plug library has a
+Router module that allows you -amongst other things- to define what to do when a
+request comes in, like so:
+
+{% highlight elixir %}
+  get "/hello" do
+    send_resp(conn, 200, "world")
+  end
+{% endhighlight %}
+
+Now, that is nice and simple, however, there is a slightly terrifying detail:
+that `conn` variable has entered our scope and was assigned a value outside of
+our control. The value is a Plug connection struct, which represents an incoming
+request and is passed around the application in a pipeline fashion to be
+transformed into a response that is then sent back.
+
+Since `get` is a macro, we need to understand macros before we can find out how
+that `conn` variable magically made it into our code.
+
+The word macro can mean a number of things, but in Elixir it is the mechanism
+that allows meta-programming. Meta-programming is when you write code whose
+purpose is to create other code, rather than executing. If you feel uncomfortable
+with this, you are having the right reaction: meta-programming is a very sharp
+tool that should only be used as a last resort, the main reason being how much
+harder this makes your code to understand.
+
+The compilation process in Elixir (or at least my very limited understanding of
+it!) involves generating an Abstract Syntax Tree (AST) from the source code,
+which is then fed to the compiler to generate bytecode for the virtual machine. The
+AST is simply a 'translation' of our source code into a format that is understood
+by the compiler. Interestingly, the reason why all these languages exist is
+because we cannot speak machine-language and the machine cannot speak human
+language (the source code), so the same 'content' (the program) has to be translated
+from human (the source code) all the way down to something that controls the
+electrons flowing through the processor.
+
+When we write a macro, we directly manipulate the AST. The major building block
+for macros is `quote` (which is itself a macro!); this converts any expression
+into its AST equivalent. This is a tuple with three elements; let's have a look:
+
+{% highlight elixir %}
+  iex()> quote do: sum(1,2,3)
```

Comment:
I would consider breaking this code snippet into 3 separate blocks with an explanation for each. It'll remove the reader having to jump back/forth